libs/providers/flagsmith/src/lib/type-factory.ts

```
@@ -0,0 +1,36 @@
+import type { FlagValue } from '@openfeature/web-sdk';
+
+export type FlagType = 'string' | 'number' | 'object' | 'boolean';
+
+/**
+ * Ret a value of the specified type based on the type parameter.
+ *
+ * @param value - The value to be converted or validated.
+ * @param type - The target type for the conversion.
+ * @returns The converted value if successful, or null if conversion fails or the type is unsupported.
+ */
+export const typeFactory = (
+  value: string | number | boolean | null | undefined,
+  type: FlagType,
+): FlagValue | undefined => {
+  if (value === null) return undefined;
+  switch (type) {
+    case 'string':
+      return value !== null && typeof value !== 'undefined' ? `${value}` : value;
+    case 'number':
+      return typeof value === 'number' ? value : parseFloat(value as string) || value;
+    case 'boolean':
+      return typeof value === 'boolean' ? value : false;
+    case 'object':
+      if (typeof value === 'string') {
+        try {
+          return JSON.parse(value);
+        } catch (error) {
+          return value;
+        }
+      }
+      return value;
+    default:
+      return value;
+  }
+};
```

For now, I've just copy-pasta'd this from [here](https://github.com/Flagsmith/js-sdk-contrib/blob/main/libs/providers/flagsmith-client/src/lib/type-factory.ts). There's probably a more elegant way to solve it, but I also ideally don't want to add overheads of maintaining more packages.

---

What's the expected behavior here if the type doesn't match the value' one ?
We are returning the value in several paths but i'm seeing in other providers a `TypeMismatchError`

cf `config-cat-web-provider.spec.ts`
```
    it('should throw TypeMismatchError if type is different than expected', () => {
      expect(() => provider.resolveBooleanEvaluation('number1', false, { targetingKey })).toThrow(TypeMismatchError);
    });
```

---

I guess the specification for this is [here](https://openfeature.dev/specification/sections/flag-evaluation#requirement-134). 

The specific requirement is this: 

> The client SHOULD guarantee the returned value of any typed flag evaluation method is of the expected type. If the value returned by the underlying provider implementation does not match the expected type, it's to be considered abnormal execution, and the supplied default value should be returned.

So, I guess, we should probably raise an exception.

---

Mmh, it's a bit ambiguous in the end. The TypeMismatchError should not happen if we follow their "the supplied default value should be returned"
As per their interface
`resolveStringEvaluation(flagKey: string, defaultValue: string, context: EvaluationContext, logger: Logger): Promise<ResolutionDetails<string>>;`

a default value should always be provided. So we shouldn't have to throw. From other providers, I can see it's doable to return something like:
```
 if (typedValue === undefined || typeof typedValue !== flagType) {
      return {
        value: defaultValue,
        reason: StandardResolutionReasons.ERROR,
        errorCode: ErrorCode.TYPE_MISMATCH,
        errorMessage: `Flag value ${flag.value} is not of type ${flagType}`,
        flagMetadata: {},
      };
    }
```
and then i'll see with open-feature team during review what's the recommended approach